### PR TITLE
refactor: simplify Deep-* types via TS 3.7 recursive types (#99)

### PR DESCRIPTION
**Pull request by lucifer1004**
All Deep-* types have been greatly simplified, including `DeepOmit`.

This PR requires TypeScript 3.7, which is a breaking change.

### CODE REVIEW

File: lib/types.ts [hunk 1]
```diff
@@ -138,9 +124,7 @@ export type MarkRequired<T, RK extends keyof T> = Exclude<T, RK> & Required<Pick
 export type MarkOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
 
 /** Convert union type to intersection #darkmagic */
-export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends ((k: infer I) => void)
-  ? I
-  : never;
+export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;
```
Comment (quezak) [inline lib/types.ts]: pls undo this formatting change, even thought the brackets and newlines are not explicitly necessary,  they did considerably improve readability of this complex expression

File: lib/types.ts [hunk 2]
```diff
@@ -158,7 +142,7 @@ export type AsyncOrSync<T> = PromiseLike<T> | T;
 // This potentially abuses compiler some inconsistencies in checking type equality for generics,
 // because normally `readonly` doesn't affect whether types are assignable.
 // @see https://stackoverflow.com/a/52473108/1815209 with comments
-type IsEqualConsideringWritability<X, Y> = (<T>() => T extends X ? 1 : 2) extends (<T>() => T extends Y ? 1 : 2)
+type IsEqualConsideringWritability<X, Y> = (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
```
Comment (quezak) [inline lib/types.ts]: same here

**Comment by quezak** [comment]
Looks good besides the one formatting comment, good timing since I wanted to hop on to this myself :)

@krzkaczor are we going to do a major release straight away, or merge this to a branch and release as `@next` first?

File: lib/types.ts (hunk 1 above)
Comment (krzkaczor) [reply lib/types.ts]: I think this is breaking change in prettier so I don't think that we can do anything about it...

File: package.json [hunk 3]
```diff
@@ -26,9 +26,9 @@
...
   "devDependencies": {
     "@codechecks/build-size-watcher": "^0.1.0",
-    "@codechecks/client": "^0.1.4",
-    "conditional-type-checks": "^1.0.1",
-    "prettier": "^1.18.2",
-    "typescript": "^3.5.2"
+    "@codechecks/client": "^0.1.10",
+    "conditional-type-checks": "^1.0.4",
+    "prettier": "^1.19.1",
+    "typescript": "^3.7.2"
```
Comment (quezak) [inline package.json]: one more thing: you need to also update TS version in the `peerDependencies` above

**Comment by krzkaczor** [comment]
@quezak I'll do a major release right after its merged.

File: lib/types.ts (hunk 1 above)
Comment (lucifer1004) [reply lib/types.ts]: Yes, the format is done by prettier.

File: package.json (hunk 3 above)
Comment (lucifer1004) [reply package.json]: Done.

File: lib/types.ts (hunk 1 above)
Comment (quezak) [reply lib/types.ts]: :( that's why I "really don't like" it, but no choice here then

Comment (krzkaczor) [reply lib/types.ts]: > OK BOOMER 💥 

😉

**Comment by krzkaczor** [comment]
Nice one, thanks!

**Pull request merged**